data_structures/bloom_filters/counting_bloom_filter/main.c

```diff
@@ -0,0 +1,79 @@
+/**
+ * @file main.c
+ * @brief Test and example code for counting bloom filter
+ * [counting bloom
+ * filter](https://en.wikipedia.org/wiki/Counting_countingBloom_filter)
+ * @details A Bloom filter is a space-efficient probabilistic data structure,
+ * that is used to test whether an element is a member of a set. False positive
+ * matches are possible, but false negatives are not – in other words, a query
+ * returns either "possibly in set" or "definitely not in set". Elements can be
+ * added to the set, and in this counting variant they can also can also be
+ * removed because the buckets use counters instead of bits. Thus an element can
+ * be removed simply by decrementing all its corrsponding buckets.  (description
+ * of bloom filter from wikipedia)
+ * @author [Eric Breyer](https://github.com/ericbreyer)
+ * @see counting_countingcountingBloom_filter.c,
+ * counting_countingcountingBloom_filter.h
+ */
+#include <assert.h>  /// assertions
+#include <stdio.h>   /// printing
+#include <stdlib.h>  /// rand
+
+#include "counting_bloom_filter.h"  /// include the structure
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test()
+{
+    printf("Constructing a new filter.\n");
+    struct CountingBloomFilter* bf = construct_CountingBloomFilter(100, .001);
+    int testNum;
+    printf("Inserting elements into the filter: ");
+    for (int i = 0; i < 10; ++i)
+    {
+        testNum = i;
+        printf("%d, ", i);
+        countingBloom_insert(bf, &testNum, sizeof testNum);
+    }
+    printf("\n\nPrinting Stats:\n");
+    countingBloom_printStats(bf);
+    printf("\n");
+    testNum = 5;
+    printf("Checking %d's membership:\n", testNum);
+    enum bloomResponse res =
+        countingBloom_contains(bf, &testNum, sizeof testNum);
+    assert(res == PROBABLY_PRESENT);
+    printf("%d is %s.\n\n", testNum,
+           res == PROBABLY_PRESENT ? "probably present"
+                                   : "definitely not present");
+    printf("Removing %d from the filter\n", testNum);
+    countingBloom_remove(bf, &testNum, sizeof testNum);
+    printf("Checking %d's membership:\n", testNum);
+    res = countingBloom_contains(bf, &testNum, sizeof testNum);
+    assert(res == DEFINITELY_NOT_PRESENT);
```

Comment:
Well, this assertion failed when I tried to insert and remove INT_MAX, probably just a lucky false positive though, since INT_MIN worked just fine

Reply:
Seems to work okay for me when I try it, probably just a false positive